### PR TITLE
Add per mesh shader option support when mesh instancing is enabled.

### PR DESCRIPTION
**Pull request by VickyAtAZ**
[PerMeshShaderOption.zip](https://github.com/carbonated-dev/o3de/files/14594863/PerMeshShaderOption.zip)
## What does this PR do?

When mesh instancing is enabled, r_enablePerMeshShaderOptionFlags was ignored previously. 
With the change in this PR, when r_enablePerMeshShaderOptionFlags is enabled, the per mesh shader options can be applied the instance group's shared MeshDrawPacket. If all the ModelDataInstances which use the same instance group are using same shader option value, then we may apply the mesh shader options to the instance group's draw packet. 

## How was this PR tested?

Editor with grubber project (and the updated shader variant lists: https://github.com/carbonated-dev/o3de-gruber/pull/572). 
r_enablePerMeshShaderOptionFlags set to true
r_directionalShadowFilteringMethod set to 1
r_directionalShadowFilteringSampleCountMode set to 0

### CODE REVIEW

File: Gems/Atom/Feature/Common/Code/Source/Mesh/MeshFeatureProcessor.cpp [hunk 1]
```diff
@@ -836,87 +850,106 @@ namespace AZ
...
-                        // Per mesh shader option flags have changed, so rebuild the draw packet with the new shader options.
-                        for (RPI::MeshDrawPacketList& drawPacketList : modelHandle.m_drawPacketListsByLod)
+
+                        if (shaderOptionFlagsChanged)
                         {
-                            for (RPI::MeshDrawPacket& drawPacket : drawPacketList)
-                            {
-                                m_flagRegistry->VisitTags(
-                                    [&](AZ::Name shaderOption, FlagRegistry::TagType tag)
+                            // Set shader optioins here
```
Comment (moudgils) [inline Gems/Atom/Feature/Common/Code/Source/Mesh/MeshFeatureProcessor.cpp]:
```suggestion
                            // Set shader options here
```

File: Gems/Atom/Feature/Common/Code/Source/Mesh/MeshFeatureProcessor.cpp [hunk 2]
```diff
@@ -836,87 +850,106 @@ namespace AZ
...
+                                });
+                            instanceGroupDataIter->UpdateDrawPacket(*GetParentScene(), true);
 
-                        // [GHI-13619]
-                        // Update the draw packets on the cullable, since we just set a shader item.
-                        // BuildCullable is a bit overkill here, this could be reduced to just updating the drawPacket specific info
-                        // It's also going to cause m_cullableNeedsUpdate to be set, which will execute next frame, which we don't need
-                        modelHandle.BuildCullable();
+                            // Note, we don't need to call CacheRootConstantInterval() here because the root constant layout won't change
+                            // when switch shader variants.
```
Comment (moudgils) [inline Gems/Atom/Feature/Common/Code/Source/Mesh/MeshFeatureProcessor.cpp]:
```suggestion
                            // when we switch shader variants.
```

File: Gems/Atom/Feature/Common/Code/Source/Mesh/MeshInstanceGroupList.h [hunk 3]
```diff
@@ -42,16 +44,39 @@ namespace AZ::Render
         // or store it with the data for each individual instance
         MeshInstanceGroupKey m_key;
 
-        // Event to signal to any data instance that refers to this InstanceGroup that the draw packet has been updated
-        AZ::Event<> m_updateDrawPacketEvent;
-        // Connecting to an AZ::Event is not thread safe, so we need to protect that with a lock
+        // A list of ModelDataInstances which are referencing this instace group
```
Comment (moudgils) [inline Gems/Atom/Feature/Common/Code/Source/Mesh/MeshInstanceGroupList.h]:
```suggestion
        // A list of ModelDataInstances which are referencing this instance group
```

File: Gems/Atom/Feature/Common/Code/Source/Mesh/MeshInstanceGroupList.h [hunk 4]
```diff
@@ -42,16 +44,39 @@ namespace AZ::Render
...
+        // If all the ModelDataInstances which use this group are using same shader option value, then we may apply the mesh shader options to the draw packet.
+        // These are two variables which are used to indicate which shader option should be applied and which value is applied.
+        // combined shader options from any ModelDataInstance which use this group
+        uint32_t m_shaderOptionFlags = 0;
+
+        // flag shader option flags which are in use
+        uint32_t m_shaderOptionFlagMask = 0;
+
+        // Update mesh draw packet
+        // Return true the DrawPacket was rebuilt
```
Comment (moudgils) [inline Gems/Atom/Feature/Common/Code/Source/Mesh/MeshInstanceGroupList.h]:
```suggestion
        // Return true if DrawPacket was rebuilt
```

File: Gems/Atom/Feature/Common/Code/Source/Mesh/MeshInstanceGroupList.h [hunk 5]
```diff
@@ -42,16 +44,39 @@ namespace AZ::Render
...
         AZStd::mutex m_eventLock;
 
         // Enable draw motion or not. Set to true if any of mesh instance use this group has the same flag set in their ModelDataInstance
         bool m_isDrawMotion = false;
 
         // If the group is transparent, sort depth in reverse
         bool m_isTransparent = false;
+
+        // For per-mesh shader options
+        // If all the ModelDataInstances which use this group are using same shader option value, then we may apply the mesh shader options to the draw packet.
```
Comment (moudgils) [inline Gems/Atom/Feature/Common/Code/Source/Mesh/MeshInstanceGroupList.h]:
```suggestion
        // If all the ModelDataInstances within this group are using the same shader option value, then we can apply the mesh shader options to the draw packet.
```

File: Gems/Atom/Feature/Common/Code/Source/Mesh/MeshInstanceGroupList.cpp [hunk 6]
```diff
@@ -8,10 +8,64 @@
...
+        uint32_t newShaderOptionFlags = m_shaderOptionFlags;
+
+        if (m_associatedInstances.size() > 0)
+        {
+            newShaderOptionFlags = (*m_associatedInstances.begin())->GetCullable().m_shaderOptionFlags;
+        }
+
+        for (auto modelDataInstance : m_associatedInstances)
+        {
+            newShaderOptionFlagMask = ~(newShaderOptionFlags ^ modelDataInstance->GetCullable().m_shaderOptionFlags);
```
Comment (moudgils) [inline Gems/Atom/Feature/Common/Code/Source/Mesh/MeshInstanceGroupList.cpp:44]: Lets add a comment what we are trying to do for this line.

File: Gems/Atom/Feature/Common/Code/Include/Atom/Feature/Mesh/MeshFeatureProcessor.h [hunk 7]
```diff
@@ -132,11 +134,6 @@ namespace AZ
             // which are turned into instance draw calls after culling
             AZStd::vector<PostCullingInstanceDataList> m_postCullingInstanceDataByLod;
```
Comment (moudgils) [inline Gems/Atom/Feature/Common/Code/Include/Atom/Feature/Mesh/MeshFeatureProcessor.h:136]: Do we not need m_updateDrawPacketEventHandlersByLod anymore?

Comment (VickyAtAZ) [reply Gems/Atom/Feature/Common/Code/Include/Atom/Feature/Mesh/MeshFeatureProcessor.h:136]: Correct. we use the m_associateIntances to replace the event/handlers

**Review by moudgils — APPROVED** [review]

**Pull request merged**